src/util/threadpool.rs

```diff
@@ -0,0 +1,610 @@
+// Copyright 2017 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::usize;
+use std::sync::{Arc, Mutex};
+use std::sync::mpsc::{channel, Sender, Receiver};
+use std::thread::{Builder, JoinHandle};
+use std::boxed::FnBox;
+use std::collections::{BinaryHeap, HashMap, VecDeque};
+use std::cmp::Ordering;
+use std::hash::Hash;
+use std::marker::PhantomData;
+
+pub struct Task<T> {
+    // The task's number in the pool.Each task has a unique number,
+    // and it's always bigger than preceding ones.
+    id: u64,
+    // the task's group_id.
+    group_id: T,
+    task: Box<FnBox() + Send>,
+}
+
+impl<T> Task<T> {
+    fn new<F>(id: u64, group_id: T, job: F) -> Task<T>
+        where F: FnOnce() + Send + 'static
+    {
+        Task {
+            id: id,
+            group_id: group_id,
+            task: Box::new(job),
+        }
+    }
+}
+
+impl<T> Ord for Task<T> {
+    fn cmp(&self, right: &Task<T>) -> Ordering {
+        self.id.cmp(&right.id).reverse()
```

Comment:
We reverse ordering here since the `heap` would pops the greatest item first while we need to pop the task with smallest `id` first. @andelf